### PR TITLE
(PDB-1561) split command api

### PR DESCRIPTION
**Pull request by ajroetker**
Split out Command API into it's own TK service.

### CODE REVIEW

**Comment by ajroetker** [comment]
@mullr wanna take another look?

File: src/puppetlabs/puppetdb/cli/export.clj [hunk 1]
```diff
@@ -272,7 +272,7 @@
 (defn- main
   [& args]
   (let [[{:keys [outfile host port] :as opts} _] (validate-cli! args)
-        src {:protocol "http" :host host :port port :prefix "/pdb/query"}
+        src {:protocol "http" :host host :port port :prefix "/pdb/query" :version :v4}
```
Comment (mullr) [inline src/puppetlabs/puppetdb/cli/export.clj]: /pdb/cmd, :version :v1 here?

Comment (mullr) [inline src/puppetlabs/puppetdb/cli/export.clj]: Oh, this actually *is* being used for query, n/m.

**Comment by kbarber** [comment]
Do we want to just have this at /pdb/cmd/v1 ... this means that later on we can't add additional namespaces I guess.

File: src/puppetlabs/puppetdb/http/command.clj [hunk 2]
```diff
@@ -19,11 +22,37 @@
...
-  "Function validating the request then submitting a command"
-  [version]
-  (-> enqueue-command
-    mid/verify-accepts-json
-    mid/verify-checksum
-    (mid/validate-query-params {:optional ["checksum"]})
-    mid/payload-to-body-string
-    (mid/verify-content-type ["application/json"])))
+  [{:keys [authorizer] :as globals}]
+  (let [{{:keys [connection endpoint]} :command-mq} globals
```
Comment (mullr) [inline src/puppetlabs/puppetdb/http/command.clj]: These two destructurings seem like they should be squished together.

File: test/puppetlabs/puppetdb/cli/services_test.clj [hunk 3]
```diff
@@ -37,26 +38,6 @@
           (is (string? (f {:ssl-client-cn "badguy"})))
           (is (= 1 (count (logs-matching #"^badguy rejected by certificate whitelist " @logz)))))))))
 
-(deftest url-prefix-test
```
Comment (mullr) [inline test/puppetlabs/puppetdb/cli/services_test.clj]: good riddance!

**Comment by mullr** [comment]
generally lgtm, just the one comment. w.r.t. ken's comment, we talked about this today and I think agreed that the current approach is ok.

File: src/puppetlabs/puppetdb/http/command.clj (hunk 2 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/command.clj]: Either way you're going to end up needing another line to get the connection and endpoint out of globals, but I can squish command-mq into the above line

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Comment by ajroetker** [comment]
@mullr fixed for your comment!

**Comment by mullr** [comment]
cool :+1: here

File: documentation/api/metrics/v1/index.markdown [hunk 4]
```diff
@@ -103,7 +103,7 @@ available via the `/metrics/v1/mbeans` endpoint.
 
 Additionally, we also support the following explicit names:
 
-**Note:** The use of these explicit names is deprecated; please use, e.g., `/v4/commands` instead.
+**Note:** The use of these explicit names is deprecated; please use, e.g., `/pdb/cmd/v1` instead.
```
Comment (wkalt) [inline documentation/api/metrics/v1/index.markdown]: not related to this PR, but we should probably consider renaming this file "mbeans.markdown"

File: documentation/api/query/v4/upgrading-from-v3.markdown [hunk 5]
```diff
@@ -20,6 +20,10 @@ Each change below is marked with the corresponding release version. Changes mark
 * (3.0) The query API has been moved from `/` to `/pdb/query`, so
   http://localhost:8080/v4/version has become
   http://localhost:8080/pdb/query/v4/version.
+  
+* (3.0) The `/commands` endpoint has been moved to it's own API at `/pdb/cmd`, so
```
Comment (wkalt) [inline documentation/api/query/v4/upgrading-from-v3.markdown]: it's -> its

File: documentation/api/query/v4/upgrading-from-v3.markdown [hunk 6]
```diff
@@ -79,10 +83,10 @@ Each change below is marked with the corresponding release version. Changes mark
 * (3.0) PuppetDB's mbeans (listed at /metrics/v1/mbeans) are no longer prefixed with
   "com."
 
-#### /pdb/query/v4/commands
+#### /pdb/cmd/v1 (fomerly /v3/commands)
```
Comment (wkalt) [inline documentation/api/query/v4/upgrading-from-v3.markdown]: fomerly -> formerly

File: documentation/release_notes.markdown [hunk 7]
```diff
@@ -58,6 +58,8 @@ stable release of PuppetDB, which included an experimental v4 API.
     dashes, which will make it easier to interface with PDB using languages like
     javascript and python.
 
+* Split the `/v3/commands` endpoint into it's own API served at `/pdb/cmd/v1`.
```
Comment (wkalt) [inline documentation/release_notes.markdown]: its

File: test/puppetlabs/puppetdb/cli/import_export_roundtrip_test.clj [hunk 8]
```diff
@@ -88,18 +94,16 @@
                     (assoc :certname "foo.local"
                            :producer_timestamp (now)))
         report (:basic reports)
-        with-server #(svc-utils/puppetdb-instance
-                      (svc-utils/create-config)
-                      %)]
+        ]
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/cli/import_export_roundtrip_test.clj]: dangling

**Comment by wkalt** [comment]
a few random punctuation comments -- looks good to me otherwise. Needs a rebase too.

File: documentation/api/metrics/v1/index.markdown (hunk 4 above)
Comment (ajroetker) [inline documentation/api/metrics/v1/index.markdown]: I'll put it on my todo list

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Comment by ajroetker** [comment]
@mullr @wkalt updated and rebased

File: test/puppetlabs/puppetdb/cli/services_test.clj [hunk 9]
```diff
@@ -77,8 +58,8 @@
 
 (deftest query-via-puppdbserver-service
   (svc-utils/with-puppetdb-instance
-    (let [pdb-service (get-service svc-utils/*server* :PuppetDBServer)]
-      (submit-command pdb-service :replace-facts 4 {:certname "foo.local"
+    (let [pdb-cmd-service (get-service svc-utils/*server* :PuppetDBCommand)]
+      (submit-command pdb-cmd-service :replace-facts 4 {:certname "foo.local"
                                                     :environment "DEV"
```
Comment (senior) [inline test/puppetlabs/puppetdb/cli/services_test.clj]: Could use some Alt-q here and the next few changes

Comment (mullr) [inline test/puppetlabs/puppetdb/cli/services_test.clj]: +1 for review using emacs keybindings as verbs

File: puppet/lib/puppet/face/node/status.rb [hunk 10]
```diff
@@ -19,7 +19,7 @@
 
       args.map do |node|
         begin
-          response = Puppet::Util::Puppetdb::Http.action("/v4/nodes/#{CGI.escape(node)}") do |http_instance, path|
+          response = Puppet::Util::Puppetdb::Http.action("/pdb/query/v4/nodes/#{CGI.escape(node)}") do |http_instance, path|
```
Comment (senior) [inline puppet/lib/puppet/face/node/status.rb]: I'm a little surprised to see these kinds of changes in this PR. The switch to /pdb/query was made in another PR, right?

**Comment by senior** [comment]
The code looks good to me, just a few small comments.

**Comment by wkalt** [comment]
functionally tested this. looks good to me.

File: puppet/lib/puppet/face/node/status.rb (hunk 10 above)
Comment (wkalt) [inline puppet/lib/puppet/face/node/status.rb]: I think it's just that it's now relative to "" instead of "/pdb/query".

Comment (ajroetker) [inline puppet/lib/puppet/face/node/status.rb]: Before we could just set a top-level prefix but now we have two prefixes the terminus needs to know about

**Comment by ajroetker** [comment]
fixed the whitespace issue @senior

**Comment by ajroetker** [comment]
@pljenkinsro retest this please

**Pull request merged**